File: extension/sidepanel/search_provider.js
class SearchProvider {
  async search(query, apiKey) {
    if (!apiKey) {
      throw new Error("Missing Tavily API Key. Please add it in Options.");
    }
    
    try {
      const response = await fetch("https://api.tavily.com/search", {
        method: "POST",
        headers: {
          "Content-Type": "application/json"
        },
        body: JSON.stringify({
          api_key: apiKey,
          query: query,
          search_depth: "advanced",
          include_answer: false,
          include_images: false,
          include_raw_content: false,
          max_results: 5
        })
      });

      if (!response.ok) {
        throw new Error(`Tavily API failed: ${response.statusText}`);
      }

      const data = await response.json();
      return data.results || [];
    } catch (e) {
      console.error("SearchProvider error:", e);
      return [];
    }
  }
}

window.SearchProvider = SearchProvider;

File: backend/providers/search_provider.py
import abc
import httpx
from backend.core.config import settings

class BaseSearchProvider(abc.ABC):
    @abc.abstractmethod
    async def search(self, query: str) -> list:
        """
        Search for evidence matching the query.
        Returns a list of dicts with 'title', 'url', 'content', and 'raw_source' fields.
        """
        pass

class TavilySearchProvider(BaseSearchProvider):
    def __init__(self):
        self.api_key = settings.TAVILY_API_KEY
        self.is_mock = "mock" in self.api_key.lower() or not self.api_key

    async def search(self, query: str) -> list:
        url = "https://api.tavily.com/search"
        payload = {
            "api_key": self.api_key,
            "query": query,
            "search_depth": "advanced"
        }
        
        async with httpx.AsyncClient() as client:
            try:
                response = await client.post(url, json=payload, timeout=10.0)
                if response.status_code == 200:
                    data = response.json()
                    return data.get("results", [])
                else:
                    print(f"Tavily search API failed with status {response.status_code}: {response.text}")
                    return []
            except Exception as e:
                print(f"Tavily search error: {str(e)}.")
                return []
